### PR TITLE
Issue #2

### PR DESCRIPTION
**Pull request by davilag**
Create basic project structure including a HTTP client to send requests from. 
This should have some generic methods for the following.

We should be extending the default Golang net/http.

- [X] GET
- [x] POST
- [X] PATCH
- [X] PUT

Project structure with `go-sf` as root.
```
go-sf/
~/http/
~/api/
```

Also added CircleCI configuration.

This solves #2

### CODE REVIEW

File: http/http.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Error building the %s request for %s: %v", method, request.URL, err)
+	}
+
+	// Fill the request with the headers
+	for k, v := range request.Headers {
+		req.Header.Set(k, v)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
```
Comment (hexnaught) [inline http/http.go]: I remember reading a while ago that it is better to not use the default client of the http package. Not sure if that is the case anymore, but may want to change it?

```Go
client := &http.Client{}
// Not sure what a best 'default' timeout would be, 10 seconds here
client.Timeout = time.Second * time.Duration(10)
resp, err := client.Do(req)
```

Comment (davilag) [reply http/http.go]: Yeah, I'll change it now: https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

>TL;DR: Go’s http package doesn’t specify request timeouts by default, allowing services to hijack your goroutines. Always specify a custom http.Client when connecting to outside services.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    steps: 
+      - checkout 
+      - restore_cache:
+          keys:
+            - go-mod-v4-{{ checksum "go.sum" }}
+
+      - run:
+          name: Run unit tests
+          command: |
+            go test ./...
```
Comment (hexnaught) [inline .circleci/config.yml]: Slightly better output on circle, gives a nice code coverage report. The `*.out` file extension is already in the `.gitignore`.

```suggestion
            go test ./... -coverprofile coverReport.out; go tool cover -func coverReport.out
```

**Review by hexnaught — COMMENTED** [review]
Do we want to go as far as having pre-commit hooks running `go mod tidy` and `go fmt ./...`?

**Comment by davilag** [comment]
> Do we want to go as far as having pre-commit hooks running `go mod tidy` and `go fmt ./...`?

I'm running `go fmt` automatically when saving any file. I don't mind to enforce it so we have a unified codebase. Maybe another issue @JokerDan

**Comment by hexnaught** [comment]
@davilag I was just thinking more, if other people are not using an editor with a plugin that automatically runs the go formatting tool.

Fine to leave it for now.

**Comment by davilag** [comment]
@JokerDan I've just created #5

**Review by hexnaught — APPROVED** [review]

File: http/http.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	client.Timeout = time.Second * time.Duration(10)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error doing %s to %s: %v", method, request.URL, err)
+	}
+	defer resp.Body.Close()
+
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading the response when doiing %s to %s: %v", method, request.URL, err)
```
Comment (Danzabar) [inline http/http.go]: errors should start with lowercase :)

File: http/http.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	// Initialising the response
+	var response Response
+	if out != nil {
+		errJSON := json.Unmarshal(respBytes, out)
```
Comment (Danzabar) [inline http/http.go]: if errJSON := json.Unmarshal(...); err != nil {
....
}

File: http/http.go (hunk 3 above)
Comment (Danzabar) [reply http/http.go]: why not use errors.New() too?

File: http/http.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Error building the %s request for %s: %v", method, request.URL, err)
+	}
+
+	// Fill the request with the headers
+	for k, v := range request.Headers {
+		req.Header.Set(k, v)
+	}
+	client := &http.Client{}
+	client.Timeout = time.Second * time.Duration(10)
```
Comment (Danzabar) [inline http/http.go]: Can just specify this in the init of http.Client{ Timeout: time.Second... }

File: http/http.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+// Put executes a PUT request against the specified URL.
+func Put(r Request, out interface{}) (*Response, error) {
+	return do("PUT", r, out)
+}
+
+func do(method string, request Request, out interface{}) (*Response, error) {
+
+	// Build the request
+	req, err := http.NewRequest(method, request.URL, bytes.NewBuffer(request.Body))
```
Comment (Danzabar) [inline http/http.go]: Hmm, why do we rebuild the request?

File: http/http.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+// Response is the type that all the HTTP requests are going to return.
+type Response struct {
+	Code    int
+	Body    interface{}
+	Headers http.Header
+}
+
+//Request is the
```
Comment (Danzabar) [inline http/http.go]: cliff hanger...

File: http/http.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+type Request struct {
```
Comment (Danzabar) [inline http/http.go]: I don't get why we need to abstract Request/Response?

File: http/http.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
+package http
```
Comment (Danzabar) [inline http/http.go]: our package has the same name as net/http :(

File: http/http_test.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
+package http
```
Comment (Danzabar) [inline http/http_test.go]: IMO we should be moving tests to a test package

**Review by Danzabar — APPROVED** [review]
Added some comments/thoughts on this, nothing to stop us merging and dealing with them later

File: http/http.go (hunk 9 above)
Comment (hexnaught) [reply http/http.go]: ^ I brought this up, thinking it would cause issues but was happy to see how things go when we start building more of the package.

Comment (davilag) [reply http/http.go]: I did some testing around this by calling our library from `main` and it was working just fine.

If I was doing

```golang
import "net/http"

...

http.Get(...)

```
was calling golang's http client and if I was doing:

```golang
import "github.com/bluetel/go-sf/http"

...

http.Get(...)

```

This was calling our http library. I can't see a massive issue since you can always aliase them right?

File: http/http.go (hunk 7 above)
Comment (davilag) [reply http/http.go]: 🙈

File: http/http.go (hunk 3 above)
Comment (davilag) [reply http/http.go]: Didn't know about it.

File: http/http.go (hunk 8 above)
Comment (davilag) [reply http/http.go]: The reason why I created `Request` was because I previously had a lot of parameters in the methods of this package (`Get`, `Post`, `do`, etc). I decided to create this struct to make that easier. True that we could probably get rid of this struct and use the `http` one, and then I have to change the package name 😜 . I thought as well we didn't want to use `net/http` and we were going to deal with all of that here.

About the `Response` one, I did it so the package takes responsibility of dealing with the serialisation and deserialisation of the response body (if you're passing an object to the method it will fill out that object, if not, it'll return raw bytes.

File: http/http.go (hunk 9 above)
Comment (Danzabar) [reply http/http.go]: Yeah I mean its still going to work, its just a little ambiguous is all - like in our http package, its hard to see which http you're using

File: http/http.go (hunk 3 above)
Comment (davilag) [reply http/http.go]: I'm using it to format the output, checking the docs looks like it's the way to go https://golang.org/pkg/errors/#example_New_errorf

Is there any reason to use `errors.New` over `ftm.Errorf`?
